### ISSUE TITLE
Migrate elasticsearch modules to extension

### ISSUE THREAD

**Issue by garethgreenaway** [salt-extensions/extension_migration#32]
modules/elasticsearch.py
returners/elasticsearch_return.py
states/elasticsearch_index_template.py
states/elasticsearch_index.py
states/elasticsearch.py

**Comment by lkubb** [comment]
Note that there is an existing extension at https://github.com/cesan3/salt-ext-elasticsearch8 that updates the modules to work with v8, maybe @cesan3 would like to help with this.

`states/grafana.py` is likely out of place here.

**Comment by cesan3** [comment]
@garethgreenaway and @lkubb.  Yeah, I could help.   However, we don't use the returners, so I didn't include them in the extension.  If I have some guidance on how to use & test the returner I'd be happy to help.

**Comment by garethgreenaway** [comment]
The graphana state was included as it has a dependency on the elasticsearch module, this dependency was auto-generated so there are likely going to be some false positives.  I'll remove it from this issue.

**Comment by garethgreenaway** [comment]
@cesan3 if your extension is more advanced than the one that is currently in the Salt core, we can deprecate that one in favor of your one.  Additionally if you wanted to move it under the Salt Extensions ORG we can do that too.

**Comment by cesan3** [comment]
Hello @garethgreenaway @nicholasmhughes  I got the new elasticsearch extension module migration including my changes for elasticsearch8.  I will deprecate saltext.elasticsearch8 in favour of saltext.elasticsearch, given that this module handles both versions  6,7 (<8) and  8+.  
I still need to test thoroughly in my environment.  The new extension is [https://github.com/cesan3/salt-ext-elasticsearch](https://github.com/cesan3/salt-ext-elasticsearch) So, you mentioned that this could be moved under the Salt Extensions ORG, I'd also like to know how to do that and how it'll be the development under the new org.

**Comment by nicholasmhughes** [comment]
@cesan3 , you're now a maintainer of https://github.com/salt-extensions/saltext-elasticsearch

Go ahead and push your code that remote and that'll be the main repo. Any further development can be in your fork, and then merged into the primary location in this org.

The Salt Extensions org provides a single location that folks can go to and look for community-developed extensions.

**Comment by cesan3** [comment]
> @cesan3 , you're now a maintainer of https://github.com/salt-extensions/saltext-elasticsearch
> 
> Go ahead and push your code that remote and that'll be the main repo. Any further development can be in your fork, and then merged into the primary location in this org.
> 
> The Salt Extensions org provides a single location that folks can go to and look for community-developed extensions.

Ok, now, I have 2 questions.
1.  The workflows depend on some permissions, so, to release I use tagpr to create the release PR so, I need 
![image](https://github.com/salt-extensions/extension_migration/assets/3269106/24a8e358-0790-4bbd-a695-d83cdca80ef1)
2. I need to add the token to be able to publish to pypi.

**Comment by cesan3** [comment]
Hmm, found some issues testing in my environment with the Lazy loader unable to load my per version utilitary modules. I'll keep working to fix the implementation

**Comment by cesan3** [comment]
So, I finally decided to use 2 different files in the module with conditional __virtualname__, much simpler that way and the fight with LazyLoader was left for another day.  I published to my repo which has the Actions enabled to release the package and publish to PyPi [https://pypi.org/project/saltext.elasticsearch/](https://pypi.org/project/saltext.elasticsearch/)  [https://github.com/cesan3/salt-ext-elasticsearch](https://github.com/cesan3/salt-ext-elasticsearch) and also pushed to the ORG repo [https://github.com/salt-extensions/saltext-elasticsearch](https://github.com/salt-extensions/saltext-elasticsearch).

**Comment by nicholasmhughes** [comment]
@cesan3 , it sounds like the new stuff is in a good spot. has the deprecation of the modules in the core been submitted?

**Comment by cesan3** [comment]
Hey @nicholasmhughes I don't have enough privileges in the repository. How can I enable actions in the ORG https://github.com/salt-extensions/saltext-elasticsearch repository? I need actions enabled to be able to release and use tagpr to open release PR.  I also need to add my PyPI secret token to be able to release and publish the package.

**Comment by nicholasmhughes** [comment]
@cesan3 , apologies for the delay in responding. we're starting to come up with a plan for consolidating the actions and creating a formal release process that can be used for all extensions. we'd love your input on how you're currently performing those functions!

**Comment by cesan3** [comment]
Hello @nicholasmhughes.  Sorry for the late response, I'm currently traveling, so limited wifi access right now.

For the elasticsearch extension module, I'm using TAGPR release process. The release creation process is triggered once the PR is merged to master/main, then, based on labels (patch, minor, major) the semantic release version is calculated.  The TAGPR will create a release PR that needs to be approved/merged to create the release, to have a bit of control.

You can check [https://github.com/cesan3/salt-ext-elasticsearch/blob/main/.github/workflows/release.yml](https://github.com/cesan3/salt-ext-elasticsearch/blob/main/.github/workflows/release.yml)

Cheers, and have a happy new year!

**Comment by nicholasmhughes** [comment]
No worries! Hope you're having a good trip @cesan3

I've implemented some "reusable workflows" housed in the `salt-extensions/central-artifacts` repository based on what the Core Team is doing with salt-describe. This should allow us to more easily keep GitHub Actions up-to-date without requiring PRs to be opened for every extension. The proposed workflow YAML can be viewed in the `central-artifacts` repository, but for posterity the Push/PR and Release pipelines look something like this:

* **Push/PR**
  * Automatic Processes (kicked off by push/PR)
    * Run pre-commit checks
    * Run test matrix for Linux, Windows, and macOS
    * Run doc build

* **Tag Creation**
  * Manual Processes
    * Create a PR to update the CHANGELOG with `towncrier`
    * Create a tag at the merged commit above for the release version using semantic versioning (i.e. `v1.0.0`)
  * Automatic Processes (kicked off by tag creation)
    * Run pre-commit checks
    * Run test matrix for Linux, Windows, and macOS
    * Run doc build
    * Build the package
    * Release the package to Test PyPi (uses the saltext-community account)
    * Create a GitHub Release for the package
    * Release the package to PyPi (uses the saltext-community account)

This is a little different than your current workflow, but I think we end up in the same place. I think the options for moving forward are:

1. Adopt the "standard" centralized workflow. All you'd really need to do is put in a PR (or I can) to throw the YAML files from `saltext-vault/.github/workflows/` into the same directory in the elasticsearch extension and then make the `saltext-community` account on PyPi an owner of the package.
2. Keep the workflow the same, but still make the `saltext-community` account on PyPi an owner of the package so you can use the organization secrets for the release.
3. I can bump your privileges in the repo so you can add your PyPi creds, but note that all organization owners will be able to see them.

Let me know how you'd like to proceed!

**Comment by cesan3** [comment]
Hello @nicholasmhughes.  I went for Option 1 and setup the workflow as suggested. I added the saltext-community user to PyPi as owner, but invitation is still pending, so, the last test to release v1.2.1 failed because of that: https://github.com/salt-extensions/saltext-elasticsearch/actions/runs/7564715861/job/20599487679.  I guess, whenever it's approved, we could successfully test the workflow.

**Comment by nicholasmhughes** [comment]
I accepted the invitation for the saltext-community PyPi user and kicked off a re-run of the failed jobs. The [publish job worked as expected](https://github.com/salt-extensions/saltext-elasticsearch/actions/runs/7564715861). The exit status failure was probably just due to the delay between the initial run and when I got around to re-running. I think we're in a good spot!
